src/services/viewport-listerner/viewport-listener.service.ts

```diff
@@ -0,0 +1,77 @@
+import { ElementRef, Injectable, OnDestroy, Renderer2 } from '@angular/core';
+import { ViewportRuler } from '@angular/cdk/scrolling';
+import { ResizeService } from '../../directives/resize/index';
+import { takeUntil } from 'rxjs/operators';
+import { fromEvent, Subject } from 'rxjs';
+
+@Injectable()
+/** Rename  */
+export class ViewportListenerService {
+
+    constructor(
+        public elementRef: ElementRef,
+        public _resizeService: ResizeService,
+        public _viewportRuler: ViewportRuler,
+        public _renderer: Renderer2) {
+    }
+
+    public createViewportListener(element: ElementRef | HTMLElement, parentElement?: ElementRef | HTMLElement): ViewportListener {
+
+        const nativeElement = element instanceof ElementRef ? element.nativeElement : element;
+
+        const nativeElementParent = parentElement instanceof ElementRef ? parentElement.nativeElement : element;
+
+        return new ViewportListener(nativeElement, nativeElementParent, this._resizeService, this._viewportRuler, this._renderer);
+    }
+
+}
+
+export class ViewportListener implements OnDestroy {
+
+    /** Allow subscribing to state changes */
+    change$ = new Subject<ViewportDirection>();
+
+    /** Store the last known position and size */
+    private _rect: ClientRect;
+
+    private _onDestroy = new Subject<void>();
+
+    constructor(private _element: HTMLElement,
+                private _elementParent: HTMLElement,
+                private _resizeService: ResizeService,
+                private _viewportRuler: ViewportRuler,
+                private _renderer: Renderer2) {
+
+        // watch for changes to the typeahead size
+        this._resizeService.addResizeListener(this._element).pipe(takeUntil(this._onDestroy))
+            .subscribe(() => {
+                this.onScrollOrResize();
+            });
+
+        // watch for changes to the typeahead position when scrolling
+        fromEvent(window, 'scroll', { passive: true }).pipe(takeUntil(this._onDestroy))
+            .subscribe(() => this.onScrollOrResize());
+
+    }
+
+    ngOnDestroy(): void {
```

Comment:
Looking at the current code, I don't think this will be invoked. `ngOnDestroy` would be called if defined on the service, but not a plain object like this. You could rename this to `destroy` and call it in the `ngOnDestroy` hook of the typeahead component. However, since the code is not really making use of the `ViewportListener` object, you could just move the implementation into the service and eliminate `ViewportListener`. In that case `ngOnDestroy` would be called automatically when the component that injected it gets destroyed.